### PR TITLE
e2e-Normalize Test Scenarios

### PR DESCRIPTION
**Pull request by itsmekumari**
@sawantpritam Please review the Transpose (Normalize) transform test scenarios

### CODE REVIEW

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature [hunk 1]
```diff
@@ -0,0 +1,105 @@
+@Normalize
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: Rename file -> BigQueryToNormalize.feature

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+Feature: Normalize transform - Verify BigQuery to BigQuery sink plugin data transfer scenarios using Normalize transform
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: -> Verify BigQuery source data transfer using Normalize transformation

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    When Select plugin: "BigQuery" from the plugins list as: "Sink"
+    Then Connect plugins: "Transpose" and "BigQuery2" to establish connection
+    Then Navigate to the properties page of plugin: "BigQuery"
+    Then Replace input plugin property: "projectId" with value: "projectId"
+    Then Enter input plugin property: "datasetProjectId" with value: "projectId"
+    Then Enter input plugin property: "referenceName" with value: "BQReferenceName"
+    Then Enter input plugin property: "dataset" with value: "dataset"
+    Then Enter input plugin property: "table" with value: "bqSourceTable"
+    Then Click on the Get Schema button
+    Then Capture the generated Output Schema
```
Comment (sawantpritam) [inline e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: No need of capturing step if you are not validating it later in preview data.
Either " add step to verify preview data columns after clicking on preview data of sink" or "remove this step and also steps to click on preview data"

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  public static void verifyPluginPropertyInvalidInlineErrorMessageColor(String propertyName) {
+    logger.info("Verify if plugin property: " + propertyName + "'s inline error message is shown in the expected color:"
+                  + " " + "#a40403");
+    String actualColor = ElementHelper.getElementColorCssProperty(NormalizeLocators.
+      locatePluginPropertyInvalidInlineError(propertyName));
+    String expectedColor = "#a40403";
+    Assert.assertEquals(expectedColor, actualColor);
+  }
+
+  @Then("Select Normalize plugin property: output schema with value: {string}")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: -> Select Normalize plugin output schema action: {string}
Update usage in Scenario as -> Select Normalize plugin output schema action: "clear" (no need of plugin property)

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * Normalize plugin step actions.
+ */
+public class NormalizeActions {
+
+  static {
+    SeleniumHelper.getPropertiesLocators(NormalizeLocators.class);
+  }
+
+  public static void selectClearOutputSchema(String clear) {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: -> selectOutputSchemaAction(String action)

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    ElementHelper.selectDropdownOption(NormalizeLocators.action,
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: -> ElementHelper.selectDropdownOption(CdfSchemaLocators.schemaActions,
                                       CdfPluginPropertiesLocators.locateDropdownListItem(action);
    WaitHelper.waitForElementToBeHidden(CdfSchemaLocators.schemaActionType(action));

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  public void selectNormalizePluginPropertyOutputSchemaWithValue(String action) throws InterruptedException {
+      NormalizeActions.selectClearOutputSchema(PluginPropertyUtils.pluginProp(action));
+    }
+
+  @Then("Enter Normalize plugin FieldsMapping {string}")
+  public void enterNormalizePluginFieldsMapping(String jsonFieldsMapping) {
+    NormalizeActions.enterFieldsMapping(jsonFieldsMapping);
+  }
+
+  @Then("Enter outputSchema {string}")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: -> Enter Normalize plugin outputSchema

File: e2e-test/src/test/features/NormalizeErrorScenarios.feature [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  Scenario:Verify Normalize plugin validation errors for invalid output schema
+    Given Open Datafusion Project to configure pipeline
+    When Expand Plugin group in the LHS plugins list: "Transform"
+    When Select plugin: "Transpose" from the plugins list as: "Transform"
+    Then Navigate to the properties page of plugin: "Transpose"
+    Then Enter Normalize plugin FieldsMapping "normalizeGcsInvalidFieldsMapping"
+    Then Enter Normalize plugin FieldsNormalize "invalidFieldsToBeNormalizedGCS"
+    Then Enter outputSchema "normalizeGcsInvalidOutputSchema"
+    Then Click on the Validate button
+    Then Delete the output schema added
```
Comment (sawantpritam) [inline e2e-test/src/test/features/NormalizeErrorScenarios.feature]: If you are entering invalid output schema then why do u need this step?

File: e2e-test/src/test/features/GCSNormalizeBigQuery.feature [hunk 9]
```diff
@@ -0,0 +1,55 @@
+@Normalize
```
Comment (sawantpritam) [inline e2e-test/src/test/features/GCSNormalizeBigQuery.feature]: rename file -> GCSToNormalize.feature

File: e2e-test/src/test/features/GCSNormalizeBigQuery.feature [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+Feature: Normalize transform - Verify GCS to BigQuery sink plugin data transfer scenarios using Normalize transform
```
Comment (sawantpritam) [inline e2e-test/src/test/features/GCSNormalizeBigQuery.feature]: -> Verify GCS source data transfer using Normalize transformation

File: e2e-test/src/test/features/FileNormalizeBigQuery.feature [hunk 11]
```diff
@@ -0,0 +1,54 @@
+@Normalize
+Feature: Normalize transform - Verify File to BigQuery sink plugin data transfer scenarios using Normalize transform
```
Comment (sawantpritam) [inline e2e-test/src/test/features/FileNormalizeBigQuery.feature]: -> Verify File source data transfer using Normalize transformation

File: e2e-test/src/test/features/FileNormalizeBigQuery.feature [hunk 12]
```diff
@@ -0,0 +1,54 @@
+@Normalize
```
Comment (sawantpritam) [inline e2e-test/src/test/features/FileNormalizeBigQuery.feature]: rename file -> FileToNormalize.feature

File: e2e-test/src/test/features/NormalizeErrorScenarios.feature [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  Scenario:Verify Normalize plugin validation errors for incorrect data in Normalize fields
+    Given Open Datafusion Project to configure pipeline
+    When Expand Plugin group in the LHS plugins list: "Transform"
+    When Select plugin: "Transpose" from the plugins list as: "Transform"
+    Then Navigate to the properties page of plugin: "Transpose"
+    Then Enter Normalize plugin FieldsMapping "normalizeGcsInvalidFieldsMapping"
+    Then Enter Normalize plugin FieldsNormalize "invalidFieldsToBeNormalizedGCS"
+    Then Enter outputSchema "normalizeGcsInvalidOutputSchema"
+    Then Click on the Validate button
+    Then Verify Normalize plugin in-line error message for incorrect field to normalize: "attributeType"
```
Comment (sawantpritam) [inline e2e-test/src/test/features/NormalizeErrorScenarios.feature]: This step can be added at the end of above scenario... as all other steps are same. And remove this scenario

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+ */
+public class NormalizeTransform implements CdfHelper {
+  private static final Logger logger = LoggerFactory.getLogger(CdfPluginPropertiesActions.class);
+
+  @Then("Delete the output schema added")
+  public void deleteTheOutputSchemaAdded() {
+    NormalizeActions.clickDeleteField(0);
+  }
+
+  @Then("Verify Normalize plugin in-line error message for incorrect field mapping: {string}")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: generic common method ->

Verify Normalize plugin property: {string}  in-line error message for value: {string}

Usage in scenario ->  Verify Normalize plugin property: "fieldMapping" in-line error message for value: "mappingField"

Use if else based on plugin property in method

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    String expectedColor = "#a40403";
+    Assert.assertEquals(expectedColor, actualColor);
+  }
+
+  @Then("Select Normalize plugin property: output schema with value: {string}")
+  public void selectNormalizePluginPropertyOutputSchemaWithValue(String action) throws InterruptedException {
+      NormalizeActions.selectClearOutputSchema(PluginPropertyUtils.pluginProp(action));
+    }
+
+  @Then("Enter Normalize plugin FieldsMapping {string}")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Enter Normalize plugin Fields to be Mapped {string}

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+  public void enterNormalizePluginFieldsMapping(String jsonFieldsMapping) {
+    NormalizeActions.enterFieldsMapping(jsonFieldsMapping);
+  }
+
+  @Then("Enter outputSchema {string}")
+  public void enterOutputSchema(String jsonOutputSchema) {
+    NormalizeActions.enterOutputSchema(jsonOutputSchema);
+  }
+
+  @Then("Enter Normalize plugin FieldsNormalize {string}")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Enter Normalize plugin Fields to be Normalized

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+  public static WebElement outputSchemaAddRowButton(int row) {
+    String xpath = "//*[@data-cy='Output Schema']//*[@data-cy='schema-fields-list']" +
+      "//*[@data-cy='schema-row-" + row + "']" +
+      "//button[@data-cy='schema-field-add-button']";
+    return SeleniumDriver.getDriver().findElement(By.xpath(xpath));
+  }
+
+  @FindBy(how = How.XPATH, using = "(//*[contains(@class, 'metric-records-out-label')])[2]/following-sibling::span")
+  public static WebElement recordsOutCount;
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Remove and use from framework.. this locator available in framework

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    Assert.assertEquals("Out records should match with target BigQuery table records count",
+            CdfPipelineRunAction.getCountDisplayedOnSourcePluginAsRecordsOut(), targetBQRecordsCount);
+  }
+
+  @Then("Validate OUT record count of transpose is equal to IN record count of sink")
+  public void validateOUTRecordCountOfTransposeIsEqualToINRecordCountOfSink() {
+    Assert.assertEquals((long) this.recordOutCount(), (long) this.recordInCount());
+  }
+
+  public int recordInCount() {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: remove recordInCount and recordOutCount from this file -> Add single method in action to validate

File: e2e-test/src/test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+      BigQueryClient.getSoleQueryResult(insertDataQuery);
+    } catch (NoSuchElementException e) {
+      // Insert query does not return any record.
+      // Iterator on TableResult values in getSoleQueryResult method throws NoSuchElementException
+    }
+    PluginPropertyUtils.addPluginProp("bqSourceTable", bqSourceTable);
+    BeforeActions.scenario.write("BQ Source Table " + bqSourceTable + " created successfully");
+  }
+
+    @Before(order = 1, value = "@GCS_CSV_NORMALIZE_TEST")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: indent

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+  @Then("Verify Normalize plugin in-line error message for incorrect field to normalize: {string}")
+  public void verifyNormalizePluginInLineErrorMessageForIncorrectFieldToNormalize(String fieldNormalizing) {
+    NormalizeTransform.verifyPluginPropertyInvalidInlineErrorMessage
+      ("fieldNormalizing",
+       PluginPropertyUtils.errorProp(E2ETestConstants.ERROR_MSG_NORMALIZE_INVALID_FIELDNORMALIZE)
+         .replace("FIELD_NORMALIZE", PluginPropertyUtils.pluginProp(fieldNormalizing)));
+    NormalizeTransform.verifyPluginPropertyInvalidInlineErrorMessageColor("fieldNormalizing");
+  }
+
+  public static void verifyPluginPropertyInvalidInlineErrorMessage(String propertyName,
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: -> verifyNormalizePluginPropertyInvalidInlineErrorMessageforRow(String propertyName, String row, String expectedMessage)

And Move this to action

Comment (sawantpritam) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Same comment for color method

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+  public static WebElement outputFieldName(int row) {
+    String xpath = "//*[@data-cy='schema-row-" + row + "']//input";
+    return SeleniumDriver.getDriver().findElement(By.xpath(xpath));
+  }
+
+  @FindBy(how = How.XPATH, using = "//*[@data-cy='Output Schema']//*[@data-cy='select-schema-actions-dropdown']")
+  public static WebElement action;
+
+  public static WebElement locatePluginPropertyInvalidInlineError(String propertyName) {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: locatePluginPropertyInvalidInlineErrorForRow(String propertyName, String row)

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+    return SeleniumDriver.getDriver().findElement(By.xpath("//*[@data-cy='" + propertyName + "']//div[@data-cy=" +
+                                                             "'error-text-0']"));
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: error-text-" + row +"

File: e2e-test/src/test/resources/pluginParameters.properties [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+fileMaxSplitSize=120
+regexPathFilter=.+TEST.csv
+incorrectRegexPathFilter=/tZY
+filePluginOutputFolder=file-plugin-output
+skipHeaderTrue=true
+fileSinkPathSuffix=yyyy-MM-dd-HH-mm-ss
+## FILE-PLUGIN-PROPERTIES-END
+
+## NORMALIZE-PLUGIN-PROPERTIES-START
+mappingField=test
```
Comment (sawantpritam) [inline e2e-test/src/test/resources/pluginParameters.properties]: normalizeMappingField

Comment (sawantpritam) [reply e2e-test/src/test/resources/pluginParameters.properties]: check all properties for `normalize` prefix

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+
+import java.io.IOException;
+
+/**
+ * Normalize Plugin related step design.
+ */
+public class NormalizeTransform implements CdfHelper {
+  private static final Logger logger = LoggerFactory.getLogger(CdfPluginPropertiesActions.class);
+
+  @Then("Delete the output schema added")
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: -> Delete Normalize plugin outputSchema row {int}

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+                                       CdfPluginPropertiesLocators.locateDropdownListItem(clear));
+  }
+
+  public static void clickDeleteField(int row) {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: clickOutputSchemaDeleteRowButton

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 25 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 24 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/resources/pluginParameters.properties (hunk 23 above)
Comment (itsmekumari) [reply e2e-test/src/test/resources/pluginParameters.properties]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java (hunk 22 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java (hunk 21 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 20 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 19 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 18 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java (hunk 17 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 16 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 15 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 14 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Need to discuss on this.

File: e2e-test/src/test/features/NormalizeErrorScenarios.feature (hunk 13 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/NormalizeErrorScenarios.feature]: As discussed, there is a bug hence cannot be removed.

File: e2e-test/src/test/features/FileNormalizeBigQuery.feature (hunk 12 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/FileNormalizeBigQuery.feature]: Change Done

File: e2e-test/src/test/features/FileNormalizeBigQuery.feature (hunk 11 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/FileNormalizeBigQuery.feature]: Change Done

File: e2e-test/src/test/features/GCSNormalizeBigQuery.feature (hunk 10 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/GCSNormalizeBigQuery.feature]: Change Done

File: e2e-test/src/test/features/GCSNormalizeBigQuery.feature (hunk 9 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/GCSNormalizeBigQuery.feature]: Change Done

File: e2e-test/src/test/features/NormalizeErrorScenarios.feature (hunk 8 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/NormalizeErrorScenarios.feature]: Yes, we need this step to capture the header error message.

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 7 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 6 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 5 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 4 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: Change Done

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature (hunk 3 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: Removed.

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature (hunk 2 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: Change Done

File: e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature (hunk 1 above)
Comment (itsmekumari) [reply e2e-test/src/test/features/BigQueryNormalizeBigQueryGCS.feature]: Change Done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * Normalize plugin step actions.
+ */
+public class NormalizeActions {
+  private static final Logger logger = LoggerFactory.getLogger(CdfPluginPropertiesActions.class);
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: -> NormalizeActions.class

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+      index++;
+    }
+  }
+
+  public static int getTargetRecordsCount() {
+    String incount = NormalizeLocators.targetRecordsCount.getText();
+    return Integer.parseInt(incount.replaceAll(",", ""));
+  }
+
+  public static void verifyNormalizePluginPropertyInvalidInlineErrorMessageForRow(String propertyName, int row,
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Remove `Invalid` word from name

Check for locator and color method name as well

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  public static WebElement outputSchemaDeleteRowButton(int row) {
+    String xpath = "//*[@data-cy='Output Schema']//*[@data-cy='schema-fields-list']" +
+      "//*[@data-cy='schema-row-" + row + "']" +
+      "//button[@data-cy='schema-field-remove-button']";
+    return SeleniumDriver.getDriver().findElement(By.xpath(xpath));
+  }
+
+  public static WebElement addRowButton(int row) {
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Remove unused locators

Check all

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+                                                                                  String expectedMessage) {
+    WaitHelper.waitForElementToBeDisplayed(NormalizeLocators.locatePluginPropertyInvalidInlineError(propertyName, row));
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: row - 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+    AssertionHelper.verifyElementDisplayed(NormalizeLocators.locatePluginPropertyInvalidInlineError(propertyName, row));
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: row - 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+    AssertionHelper.verifyElementContainsText(NormalizeLocators.locatePluginPropertyInvalidInlineError(propertyName,
+            row), expectedMessage);
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: row - 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  public static void verifyNormalizePluginPropertyInvalidInlineErrorMessageForColor(String propertyName, int row) {
+    logger.info("Verify if plugin property: " + propertyName + "'s inline error message is shown in the expected color:"
+            + " " + "#a40403");
+    String actualColor = ElementHelper.getElementColorCssProperty(NormalizeLocators.
+            locatePluginPropertyInvalidInlineError(propertyName, row));
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: row - 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+
+  @Then("Delete Normalize plugin outputSchema row {int}")
+  public void deleteNormalizePluginOutputSchemaRow(int row) {
+    NormalizeActions.clickOutputSchemaDeleteRowButton(row);
+  }
+
+  @Then("Verify Normalize plugin in-line error message for incorrect field mapping: {string}")
+  public void verifyNormalizePluginInLineErrorMessageForIncorrectFieldMapping(String fieldMapping) {
+    NormalizeActions.verifyNormalizePluginPropertyInvalidInlineErrorMessageForRow
+      ("fieldMapping", 0,
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: 0 -> 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 34]
```diff
@@ -0,0 +1,102 @@
...
+       PluginPropertyUtils.errorProp(E2ETestConstants.ERROR_MSG_NORMALIZE_INVALID_FIELDMAPPING)
+         .replace("FIELD_MAPPING", PluginPropertyUtils.pluginProp(fieldMapping)));
+    NormalizeActions.verifyNormalizePluginPropertyInvalidInlineErrorMessageForColor("fieldMapping", 0);
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: 0 -> 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  @Then("Verify Normalize plugin in-line error message for incorrect field to normalize: {string}")
+  public void verifyNormalizePluginInLineErrorMessageForIncorrectFieldToNormalize(String fieldNormalizing) {
+    NormalizeActions.verifyNormalizePluginPropertyInvalidInlineErrorMessageForRow
+      ("fieldNormalizing", 0,
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: 0 -> 1

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java [hunk 36]
```diff
@@ -0,0 +1,102 @@
...
+       PluginPropertyUtils.errorProp(E2ETestConstants.ERROR_MSG_NORMALIZE_INVALID_FIELDNORMALIZE)
+         .replace("FIELD_NORMALIZE", PluginPropertyUtils.pluginProp(fieldNormalizing)));
+    NormalizeActions.verifyNormalizePluginPropertyInvalidInlineErrorMessageForColor("fieldNormalizing", 0);
```
Comment (sawantpritam) [inline e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: 0 -> 1

**Review by sawantpritam — CHANGES_REQUESTED** [review]

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 26 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 27 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: Removed

File: e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java (hunk 28 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/locators/NormalizeLocators.java]: Removed

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 29 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 30 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 31 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java (hunk 32 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/actions/NormalizeActions.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 33 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 34 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 35 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: change done

File: e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java (hunk 36 above)
Comment (itsmekumari) [reply e2e-test/src/test/java/io/cdap/plugin/normalize/stepsdesign/NormalizeTransform.java]: change done

**Pull request closed without merging**